name: Report Errors

on:
  schedule:
    - cron: "0 * * * *"
  pull_request:
    paths:
      - ".github/workflows/report-errors.yml"

permissions:
  actions: write
  issues: write

jobs:
  report-errors:
    name: Report Errors
    runs-on: ubuntu-20.04
    steps:
      - id: locate-runs
        name: Locate Runs
        uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea
        with:
          script: |
            const now = new Date();
            const yesterday = new Date(now.getTime() - 24 * 60 * 60 * 1000);
            const workflowRuns = await github.paginate(github.rest.actions.listWorkflowRuns, {
              owner: context.repo.owner,
              repo: context.repo.repo,
              branch: context.repo.default_branch,
              workflow_id: "integration.yml",
              created: `>${yesterday.toISOString()}`,
            });
            const successfulRuns = workflowRuns.filter(run => run.conclusion === "success");
            if (successfulRuns.length > 0) {
              const successfulRun = successfulRuns[0];
              console.log(`Found successful run ${successfulRun.id}.`);
              return false;
            }

            const failedRuns = workflowRuns.filter(run => run.conclusion !== null && run.conclusion !== "success");
            if (failedRuns.length > 0) {
              const failedRun = failedRuns[0];
              console.log(`Found failed run ${failedRun.id}.`);
              if (failedRun.run_attempt < 5) {
                console.log("Retrying run.");
                github.request("POST /repos/{owner}/{repo}/actions/runs/{run_id}/rerun-failed-jobs", {
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  run_id: failedRun.id,
                })
                return null;
              } else {
                console.log("Run is already at maximum retries.");
                return true;
              }
            }

            console.log("No relevant runs found.");
            return false;
      - name: Create or Update Issue
        uses: chrisgavin/create-or-update-issue-action@6c33f901f0e5b25f96409e670b49814902d35fa3
        if: ${{ steps.locate-runs.outputs.result != 'null' }}
        with:
          title: Integration tests are currently failing.
          body: Integration tests on the repository are currently failing. This might be due to a change in the way Slack works.
          assignees: chrisgavin
          close: ${{ steps.locate-runs.outputs.result == 'false' }}
